twitter_api/main.py

```diff
@@ -15,7 +18,283 @@ def before_request():
     g.db = connect_db(app.config['DATABASE'])
 
 
-# implement your views here
+
+@app.route('/login', methods=['POST'])
+def login():
+    # Get user credentials from request
+    user_data = json.loads(request.data.decode('utf-8'))
+    
+    # Validate required params
+    try:
+        username = user_data['username']
+        password = user_data['password']
+    except KeyError:
+        return Response(status=400)
+    
+    # Query user table to get user data
+    query = """
+        SELECT id, username, password 
+        FROM user 
+        WHERE username=:username;
+    """
+    cursor = g.db.execute(query, {'username': username})
+    user_in_db = cursor.fetchone()
+    
+    # Validate user exists and passwords match
+    if not user_in_db:
+        return Response(status=404)
+    
+    if user_in_db[2] != hash_to_md5(password):
+        return Response(status=401)
+
+    if user_in_db[2] == hash_to_md5(password):
+        # Create access_token and query auth table to save it
+        access_token = hash_to_md5('{}{}{}'.format(SALT, username, password))
+        query = """
+            INSERT INTO auth (user_id, access_token) 
+            VALUES (:user_id, :access_token);
+        """
+        g.db.execute(
+            query, {'user_id': user_in_db[0], 'access_token': access_token})
+        g.db.commit()
+        
+        # Build object to send in response
+        json_response = json.dumps({'access_token': access_token})
+        return Response(json_response, status=201)
+    
+     
+@app.route('/logout', methods=['POST'])
+def logout():
+    # Get access_token from request.data
+    try:
+        access_token = json.loads(request.data.decode('utf-8'))['access_token']
```

Comment:
You're hardcoding the encoding 'utf-8'. A request has information about the charset used by the client.